File: Hard/closestWizzards.js
const wizzards = [
  [1, 8, 6],
  [2, 3],
  [5, 2],
  [4, 6],
  [5, 8],
  [6, 9],
  [7, 9],
  [8, 9],
  [1],
  [1, 3]
];

function findMinPath(wizzards, start = 0, goal = 9) {
  let shortestPathArr = [],
    shortestPathLen = Number.MAX_SAFE_INTEGER,
    brain = {},
    currentWizzardFriends = wizzards[start];

  for (let i = 0; i < currentWizzardFriends.length; i++) {
    let currentShortestPathArr = [0];
    let currentShortestPathLen = 0;
    _traverse(currentWizzardFriends[i], 0, currentShortestPathArr, currentShortestPathLen);
  }

  function _traverse(currentWizzard, previousWizzard, currentShortestPathArr, currentShortestPathLen) {
    currentShortestPathArr.push(currentWizzard);
    currentShortestPathLen += Math.pow(currentWizzard - previousWizzard, 2);

    if (currentWizzard === goal) {
      if (currentShortestPathLen < shortestPathLen) {
        shortestPathLen = currentShortestPathLen;
        shortestPathArr = currentShortestPathArr.slice();
        currentShortestPathArr.pop();
        return;
      }
    }

    let currentWizzardFriends = wizzards[currentWizzard];

    for (let i = 0; i < currentWizzardFriends.length; i++) {
      if (brain.hasOwnProperty(`${currentWizzard},${i}`)) {
        continue;
      }
      brain[`${currentWizzard},${i}`] = true;
      _traverse(currentWizzardFriends[i], currentWizzard, currentShortestPathArr.slice(), currentShortestPathLen);
    }
  }

  return shortestPathArr;
}

findMinPath(wizzards, 0, 9);

File: AirBnB/closestWizzards.scala
import scala.collection.mutable.HashSet
object Solution extends App {

  val wizzards1 = Array(
    Array(1, 8, 6),
    Array(2, 3),
    Array(5, 2),
    Array(4, 6),
    Array(5, 8),
    Array(6, 9),
    Array(7, 9),
    Array(8, 9),
    Array(1),
    Array(1, 3)
  )

  private var gShortestPathArr = Array[Int]()
  private var gShortestPathLen = Int.MaxValue
  private val brain = HashSet[String]()

  private def findMinPath(
      wizzards: Array[Array[Int]],
      start: Int = 0,
      goal: Int = 9
  ) = {
    val cWizardFriends = wizzards(start)

    for (i <- cWizardFriends.indices) {
      val cShortestPathArr = Array(start)
      val cShortestPathLen = 0
      traverse(
        cWizardFriends(i),
        0,
        cShortestPathArr,
        cShortestPathLen,
        goal
      )
    }

    gShortestPathArr
  }

  private def traverse(
      cWizzard: Int,
      prevWizard: Int,
      shortestPathArr: Array[Int],
      shortestPathLen: Int,
      goal: Int
  ): Unit = {
    val cShortestPathArr = shortestPathArr :+ cWizzard
    val cShortestPathLen =
      (shortestPathLen + math.pow(cWizzard - prevWizard, 2)).toInt

    if (cWizzard == goal) {
      if (cShortestPathLen < gShortestPathLen) {
        gShortestPathLen = cShortestPathLen
        gShortestPathArr = cShortestPathArr.toArray
        return ()
      }
    }

    val cWizzardFriends = wizzards1(cWizzard)

    for (i <- cWizzardFriends.indices) {
      if (!brain.contains(s"$cWizzard,$i")) {
        brain += (s"$cWizzard,$i")
        traverse(
          cWizzardFriends(i),
          cWizzard,
          cShortestPathArr,
          cShortestPathLen,
          goal
        )
      }
    }
  }

  private def stringify(obj: Any): String = {
    scala.runtime.ScalaRunTime.stringOf(obj)
  }

  println(stringify(findMinPath(wizzards1, 0, 9))) // Array(0, 1, 2, 5, 6, 7, 9)
}
